### DISCUSSION TITLE
Rock Paper Scissors tutorial program was working, but stopped. What went wrong?

### DISCUSSION THREAD

**Discussion by JamesLampert** [reach-sh/reach-lang#816] [Q&A] [answered]
Some weeks ago, I got through the Rock Paper Scissors tutorial, following Ms. Hodge's "Ten Days of Blockchain" video series. I got it to the point where "Bob" and "Alice" were in separate frontend instances, in separate terminal sessions (Day 9), and I moved it to TestNet, (Day 10) using existing, persistent accounts. Here's the relevant part of my frontend:

```
    let ctc = null;
    const deployCtc = await ask(
        'Do you want to deploy the contract (y/n)',
        yesno
        )
    
    if ( deployCtc ) {
        console.log('Deploying')
        ctc = acc.contract(backend)
        ctc.getInfo().then((info => {
            console.log(`The contract is deployed as = ${JSON.stringify(info)}`)
        }))
    } else {
        const info = await ask(
            'Please paste the contract information:',
            JSON.parse
        )
        ctc = acc.contract(backend, info)
    }

```

And it was working. Some weeks ago.

Now (or at least, last week), when I tried it, "Alice" answers "yes" on deploying the contract, but
```
            console.log(`The contract is deployed as = ${JSON.stringify(info)}`)
```
never fires, and no error message appears. I added "console.log('Deploying')" to verify that it was getting that far (it is), but we never get a contract ID, and I have no idea what is going wrong, or even what the next step is in finding out.

**Comment by jeapostrophe** [comment, marked as answer]
It will not print until after Alice provides the wager, because the program isn't actually deployed until the first message.

**Reply by JamesLampert** [reply]
I should've re-watched Day 9 of Ms. Hodge's video series. That's _exactly_ how it behaves.

And while I must have accepted it at the time, I'm having to work at wrapping my age-addled grey cells around a control flow that would produce that behavior.

If Alice's `ctc = acc.contract(backend)` and Bob's `ctc = acc.contract(backend, info)` are passing control to the Reach program, then how does it get to the console.log statement that displays the initial balance before it gets (on Alice's side) to the console.log statement that displays the contract ID? Is this a matter of threads being spawned off, without anything that I recognize as explicitly spawning them?

**Reply by jeapostrophe** [reply]
Yup. Promises are threads. Then .then is spawning a thread that waits until the info is available

**Reply by JamesLampert** [reply]
Thanks. I'm probably going to keep annoying you about this until I fully understand what's going on, but at least I'm no longer beating my head against a wall.

**Reply by JamesLampert** [reply]
Ok. Let me see if I understand the flow of control now.

For Alice, deploying the contract:
index.mjs establishes that the player is playing as Alice, gets the account mnemonic, and asks whether or not to deploy the contract.
It then logs a message that it is deploying the contract, spawns off index.rsh, and then (the ".then" clause) spawns off a thread that waits for the contract ID to become available, and logs it (and presumably terminates) as soon as it is.
Without waiting for anything in any other thread, it then displays the balance.
Then, `await part(ctc, interact)` waits for the game to conclude, before printing the balance and terminating.

Meanwhile, index.rsh gets the wager and deadline (calling back to functions in index.mjs, that ask for the wager and selects the deadline based on the network). It then publishes the wager and deadline (causing the contract ID thread to unblock and print the ID), and draws the wager into the contract.
After interacting with Bob to set up the wager, it then loops through a minimum of one round of the game, until a player wins:
First, it gets Alice's hand (by calling back to a function in index.mjs that asks for it), commits it, and publishes her commitment
Then it gets Bob's hand (the same way), publishes it, publishes Alice's hand and salt, checks the commitment, and determines the outcome.
Finally, it transfers the pot to the winner, calls frontend functions to display the outcome to both players, and terminates (causing the index.mjs thread to unblock and display Alice's ending balance).


For Bob, not deploying the contract:
index.mjs establishes that the player is playing as Bob, gets the account mnemonic, and asks whether or not to deploy the contract. It then accepts the contract ID, and spawns off index.rsh on the selected contract.
Without waiting for anything in any other thread, it then displays the balance.
Then, `await part(ctc, interact)` waits for the game to conclude, before printing the balance and terminating.

Meanwhile, index.rsh, after interacting with Alice to set up the wager and deadline, calls back to a frontend function that displays the wager and asks if Bob accepts. It then draws the wager into the contract.
Then, it loops through a minimum of one round of the game, as described above.
Finally, it transfers the pot to the winner, calls frontend functions to display the outcome to both players, and terminates (causing the index.mjs thread to unblock and display Bob's ending balance).

Am I reasonably close to describing what is actually going on?

**Reply by jeapostrophe** [reply]
Very close. The only difference is that `await part....` is actually the thing that "spawns off `index.rsh`". The `.then` clause spawned a JS thread that is waiting for the contract info to be available.

**Reply by JamesLampert** [reply]
Hmm.

I think perhaps part of the confusion I'm experiencing is that instead of functions being defined in a block above the main (as in Pascal) or below it (as in FORTRAN), I'm seeing functions being defined in the middle of everything. And I think the last time I wasn't able to run a program in some sort of debugger, and actually see (and single-step) the flow of control, I was still a student at CSU Long Beach, writing mostly in PL/I.

**Reply by jeapostrophe** [reply]
That is incorrect. The functions are only defined above and then called below. We don't use any recursive or forward references in this code.

**Reply by JamesLampert** [reply]
Pascal: Forward references are categorically prohibited unless explicitly declared (Wirth was apparently allergic to multipass compilers and linkage editors), but recursion is permitted. Main program is always at the bottom, and cannot be interrupted for functions or procedures.

FORTRAN: Forward references (and references to stuff in separately compiled modules) are categorically permitted, and references from the main are always forward references, but recursion is categorically prohibited (until FORTRAN 90, which I only just now even heard about). Main program is always at the top, and cannot be interrupted for functions or procedures.

JavaScript (and maybe Reach) looks like functions can be defined at any point before they are called: in R/P/S, we see interact.informTimeout defined between displaying the balance and asking for the wager, and interact.getHand and interact.seeOutcome defined between displaying the balance and awaiting the Reach program.

**Reply by jeapostrophe** [reply]
If you are talking about `interact`, that is JavaScript. If you are talking about this file --- https://github.com/reach-sh/reach-lang/blob/master/examples/rps-8-interact/index.mjs --- no function is defined after code that calls it and there are no forward references. The main program is at the bottom. We write JavaScript like Pascal.
